first attempt of sd-jwt vcdm (#134)
openid/oid4vc-haip: openid4vc-high-assurance-interoperability-profile-sd-jwt-vc-1_0.md

* The following JWT Claims MUST be supported in addition to the requirements in Section 3.2.2.2 in [@!I-D.ietf-oauth-sd-jwt-vc]:

| Claim | SD-JWT as issued by the Issuer | Normative Definition |
|:--- |:--- |:--- |
|iat |MUST |[@!RFC7519], Section 4.1.6 |
|cnf|	MUST|	[@!RFC7800]|
|exp | SHOULD (at the discretion of the issuer) | [@!RFC7519], Section 4.1.4 |
|status|SHOULD (at the discretion of the issuer)| [@!I-D.ietf-oauth-status-list]|

* The Issuer MUST NOT make any of the JWT Claims in the table above to be selectively disclosable, so that they are always present in the SD-JWT VC presented by the Holder.
* It is at the discretion of the Issuer whether to use `exp` claim and/or a `status` claim to express the validity period of an SD-JWT VC. The Issuer MUST use one of the mechanisms. The wallet and the verifier MUST support both mechanisms.
* The `iss` claim MUST be an HTTPS URL. The `iss` value is used to obtain Issuer’s signing key as defined in (#issuer-key-resolution).
* The `cnf` claim [@!RFC7800] MUST include the JSON Web Key [@!RFC7517] in the `jwk` sub claim.

> Note: Currently this profile only supports presentation of credentials with cryptographic Holder Binding: the holder's signature is required to proof the credential is presented by the holder it was issued to. This profile might support claim-based and biometrics-based holder binding once OpenID for Verifiable Credentials adds support for other forms of Holder Binding. See https://bitbucket.org/openid/connect/issues/1537/presenting-vc-without-a-vp-using-openid4vp

Reviewer: Should this section be removed? Feels like it's using old POV and references an old issue

A Credential Format Profile for Credentials complying with IETF SD-JWT VCs [@!I-D.ietf-oauth-sd-jwt-vc] is defined in Annex A.3 of [@!OIDF.OID4VCI] and Annex A.4 of [@!OIDF.OID4VP].

> ## SD-JWT VCDM

Reviewer 2: We should clearly say something about the goals of SD-JWT VCDM and who can consume it. For example, we should define what of the following scenarios are possible:
- SD-JWT VC processor without JSON-LD processing
- SD-JWT VC processor with JSON-LD processing
- W3C VCDM 2.0 processor (which requires JSON-LD), with securing mechanism verification. This might be an issue because of the external context definitions will be required, and because SD-JWT VC has to be defined as a W3C VCDM 2.0 securing mechanism.
- W3C VCDM 2.0 processor without securing mechanism verification; just do SD-JWT VC verification. This will also require external context definitions for all public/private JWT claims which are not defined in the W3C VCDM 2.0 context already. 
- any other options? 

After this PR got merged: https://github.com/w3c/vc-data-model/pull/1520, VCDM 2.0 requires all claims to be defined explicitly in dedicated `@context` definitions referenced by the `@context` array. This is required because all VCDM 2.0 credentials have to be JSON-LD compliant. Before this PR, this was not strictly required (although kind of recommended). Not having these properly defined in these context definitions would break VCDM 2.0 processors. The VCDM 2.0 context does have definitions for all the standard SD-JWT terms but does not have definitions of SD-JWT VC claims such as `status`, `schema` etc. Previously all these claims got added the default context as `issuer-dependent` terms via the `"@vocab": "https://www.w3.org/ns/credentials/issuer-dependent#"` statement in the W3C VCDM 2.0 context.

Reviewer 2: It is possible to add a default vocab in HAIP but it should be noted that this MUST NOT be done if Data Integrity Proofs are used. I'm not sure yet if this is a good idea though.

However, we would need to use a different IRI for the value of `@vocab` since https://www.w3.org/ns/credentials/issuer-dependent#" does not go anywhere.

SD-JWT VCDM MAY be used. SD-JWT VCDM is a data model that follows [@!I-D.ietf-oauth-sd-jwt-vc], but is compatible with the [@!W3C.VCDM2.0].

> The following is a non-normative example of an unsecured payload of an SD-JWT VCDM, that is built using the example of unsecured payload in Section 3.3 of [@!I-D.ietf-oauth-sd-jwt-vc]:

Reviewer 3: Since SD-JWT VC DM is not a specification, I would say "in accordance with the SD-JWT VC DM proposal, " or similar.

Author: the point is to define SD-JWT VCDM in HAIP, here, from scratch

Reviewer 4: Should the authors contribute?

Reviewer 4: Why is the sd-jwt vcdm being moved here? Was there an agreement with the authors?

```json
{
  "vct": "https://credentials.example.com/identity_credential",

//W3C VCDM 2.0 compliant claims
  "type": ["VerifiableCredential", "https://credentials.example.com/identity_credential"],
  "@context": ["https://www.w3.org/ns/credentials/v2"],
  "issuer": "https://example.com/issuer",
  "credentialSubject": {
    "given_name": "John",
    "family_name": "Doe",
    "email": "johndoe@example.com",
    "phone_number": "+1-202-555-0101",
    "address": {
      "street_address": "123 Main St",
      "locality": "Anytown",
      "region": "Anystate",
      "country": "US"
    },
    "birthdate": "1940-01-01",
    "is_over_18": true,
    "is_over_21": true,
    "is_over_65": true
  }
}
```

The following is a non-normative example of how an unsecured payload of an SD-JWT VCDM above can be used for the SD-JWT:

```json
{
  "_sd": [
    "vwUhdFvpylx9Sqi2YNBV1dfVK6lCbhXvkH0nThfKFT0"
  ],
  "iss": "https://example.com/issuer",

>   "issuer": "https://example.com/issuer",

Reviewer 5: Although the jwt encoding for w3c vcs was definitely a lot simpler than the JSON-LD counterpart, the thing i disliked about it the most was the duplication of values and the issues it caused in validation. Every duplicated parameter introduces a chance of inconsistent data in the credential. So i agree with @awoie here

  "iat": 1683000000,
  "exp": 1883000000,
  "vct": "https://credentials.example.com/identity_credential",
  "type": ["VerifiableCredential", "https://credentials.example.com/identity_credential"],
  "@context": ["https://www.w3.org/ns/credentials/v2"],
  "_sd_alg": "sha-256",
  "cnf": {
    "jwk": {
      "kty": "EC",
      "crv": "P-256",
      "x": "TCAER19Zvu3OHF4j4W4vfSVoHIP1ILilDls7vCeGemc",
      "y": "ZxjiWWbZMQGHVWKVQ4hbSIirsVfuecCE6t4jT9F2HZQ"
    }
  }
}
```

> * One of the values in `type` array MUST equal the value in `vct` claim.
> * The value of `issuer` clam MUST equal the value in `iss` claim.
> 
> For the following, mechanisms defined in [@!I-D.ietf-oauth-sd-jwt-vc] are used:
> 
> * Status management (e.g., revoked, suspended, etc.)
> * Schema
> * Display information

Reviewer 2: This section needs to be more precise. The following are just my first comments:
```suggestion
* One of the values in `type` array MUST equal the value in `vct` claim.
* The value of `issuer` clam MUST equal the value in `iss` claim.

The following mechanisms defined in [@!I-D.ietf-oauth-sd-jwt-vc] MUST be used instead of their W3C VCDM 2.0 counter part:

* Status management (e.g., revoked, suspended, etc.)
* Schema
* Display information
* Validity information
```

    "family_name": "Doe",
    "email": "johndoe@example.com",
    "phone_number": "+1-202-555-0101",
    "address": {
      "street_address": "123 Main St",
      "locality": "Anytown",
      "region": "Anystate",
      "country": "US"
    },
    "birthdate": "1940-01-01",
    "is_over_18": true,
    "is_over_21": true,
    "is_over_65": true
  }
}
```

The following is a non-normative example of how an unsecured payload of an SD-JWT VCDM above can be used for the SD-JWT:

```json
{
  "_sd": [
    "vwUhdFvpylx9Sqi2YNBV1dfVK6lCbhXvkH0nThfKFT0"
  ],
  "iss": "https://example.com/issuer",
  "issuer": "https://example.com/issuer",
  "iat": 1683000000,
  "exp": 1883000000,
  "vct": "https://credentials.example.com/identity_credential",
  "type": ["VerifiableCredential", "https://credentials.example.com/identity_credential"],
  "@context": ["https://www.w3.org/ns/credentials/v2"],
  "_sd_alg": "sha-256",
  "cnf": {
    "jwk": {
      "kty": "EC",
      "crv": "P-256",
      "x": "TCAER19Zvu3OHF4j4W4vfSVoHIP1ILilDls7vCeGemc",
      "y": "ZxjiWWbZMQGHVWKVQ4hbSIirsVfuecCE6t4jT9F2HZQ"
    }
  }
}
```

* One of the values in `type` array MUST equal the value in `vct` claim.

> * The value of `issuer` clam MUST equal the value in `iss` claim.

Reviewer: ```suggestion
* The value of `issuer` claim MUST equal the value in `iss` claim.
```

The following mechanisms defined in [@!I-D.ietf-oauth-sd-jwt-vc] MUST be used instead of their W3C VCDM 2.0 counter part:

* Status management (e.g., revoked, suspended, etc.)
* Schema
* Display information
* Validity information

# Crypto Suites

Issuers, holders and verifiers MUST support P-256 (secp256r1) as a key type with ES256 JWT algorithm for signing and signature validation whenever this profiles requires to do so:

* SD-JWT VC
* Wallet Instance Attestation
* DPoP

> * HB JWT

Reviewer: ```suggestion
* Key proof JWT
```

A Credential Format Profile for Credentials complying with IETF SD-JWT VCs [@!I-D.ietf-oauth-sd-jwt-vc] is defined in Annex A.3 of [@!OIDF.OID4VCI] and Annex A.4 of [@!OIDF.OID4VP].

## SD-JWT VCDM

SD-JWT VCDM is a data model that follows [@!I-D.ietf-oauth-sd-jwt-vc], but is compatible with the [@!W3C.VCDM2.0]. SD-JWT VCDM MAY be used.

The following is a non-normative example of an unsecured payload of an SD-JWT VCDM, that is built using the example of unsecured payload in Section 3.3 of [@!I-D.ietf-oauth-sd-jwt-vc]:

```json
{
  "vct": "https://credentials.example.com/identity_credential",

  //W3C VCDM 2.0 compliant claims

>   "type": ["VerifiableCredential", "https://credentials.example.com/identity_credential"],

Reviewer 2: I would remove this duplication since it is not required.
```suggestion
  "type": ["VerifiableCredential"],
```
